semantic/src/main/kotlin/tools/samt/semantic/SemanticModel.kt

```diff
@@ -0,0 +1,597 @@
+                        }
+                    }
+
+                    controller.createContext(expression.location.source).error {
+                        message("Type '${expression.name}' could not be resolved")
+                        highlight("unresolved type", expression.location)
+                    }
+                }
+
+                is CallExpressionNode -> {
+                    val baseType = resolveExpression(expression.base)
+
+                    fun createRange(expression: RangeExpressionNode): ResolvedTypeReference.Constraint.Range {
+                        fun resolveSide(expressionNode: ExpressionNode): Number? = when (expressionNode) {
+                            is NumberNode -> expressionNode.value
+                            is WildcardNode -> null
+                            else -> {
+                                controller.createContext(expressionNode.location.source).error {
+                                    message("Range constraint argument must be a valid number range")
+                                    highlight("invalid constraint", expressionNode.location)
+                                    help("A valid constraint would be range(1..10.5) or range(1..*)")
+                                }
+                                null
+                            }
+                        }
+                        return ResolvedTypeReference.Constraint.Range(
+                            lowerBound = resolveSide(expression.left),
+                            upperBound = resolveSide(expression.right),
+                        )
+                    }
+
+                    fun createSize(expression: RangeExpressionNode): ResolvedTypeReference.Constraint.Size {
+                        fun resolveSide(expressionNode: ExpressionNode): Long? = when (expressionNode) {
+                            is IntegerNode -> expressionNode.value
+                            is WildcardNode -> null
+                            else -> {
+                                controller.createContext(expressionNode.location.source).error {
+                                    message("Range constraint argument must be a valid integer range")
+                                    highlight("invalid constraint", expressionNode.location)
+                                    help("A valid constraint would be size(1..10) or size(1..*)")
+                                }
+                                null
+                            }
+                        }
+                        return ResolvedTypeReference.Constraint.Size(
+                            lowerBound = resolveSide(expression.left),
+                            upperBound = resolveSide(expression.right),
+                        )
+                    }
+
+                    fun createPattern(expression: StringNode): ResolvedTypeReference.Constraint.Pattern {
+                        // We should probably validate the pattern here
```

Comment:
> We should probably validate the pattern here

yes

Reply:
It's actually not trivial to verify the pattern is reasonable, given we want to target different programming languages we would have to decide on a limited regex feature set to support different kind of [flavors](https://gist.github.com/CMCDragonkai/6c933f4a7d713ef712145c5eb94a1816). I left the comment and created a ticket to look into what to do in the future

Reply:
Possible solution: Do not verify and blindly pass to target language.